R/scale-expansion.r

```diff
@@ -0,0 +1,223 @@
+  # upper range limits, and then combine them into one vector
+  lower <- expand_range(limits, expand[1], expand[2])[1]
+  upper <- expand_range(limits, expand[3], expand[4])[2]
+  c(lower, upper)
+}
+
+#' Calculate the default expansion for a scale
+#'
+#' @param scale A position scale (e.g., [scale_x_continuous()] or [scale_x_discrete()])
+#' @param discrete,continuous Default scale expansion factors for
+#'   discrete and continuous scales, respectively.
+#' @param expand Should any expansion be applied?
+#'
+#' @return One of `discrete`, `continuous`, or `scale$expand`
+#' @noRd
+#'
+expand_default <- function(scale, discrete = expand_scale(add = 0.6),
+                           continuous = expand_scale(mult = 0.05), expand = TRUE) {
+  if (!expand) {
+    return(expand_scale(0, 0))
+  }
+
+  scale$expand %|W|% if (scale$is_discrete()) discrete else continuous
+}
+
+#' Expand limits in (possibly) transformed space
+#'
+#' These functions calculate the continuous range in coordinate space
+#' and in scale space. Usually these can be calculated from
+#' eachother using the coordinate system transformation, except
+#' when transforming and expanding the scale limits results in values outside
+#' the domain of the transformation (e.g., a lower limit of 0 with a square root
+#' transformation).
+#'
+#' @param scale A position scale (see [scale_x_continuous()] and [scale_x_discrete()])
+#' @param limits The initial scale limits, in scale-transformed space.
+#' @param coord_limits The user-provided limits in scale-transformed space,
+#'   which may include one more more NA values, in which case those limits
+#'   will fall back to the `limits`.
+#' @param expansion An expansion generated by [expand_scale()] or [expand_default()].
+#' @param trans The coordinate system transformation.
+#'
+#' @return A list with components `continuous_range`, which is the
+#'   expanded range in scale-transformed space, and `continuous_range_coord`,
+#'   which is the expanded range in coordinate-transformed space.
+#'
+#' @noRd
+#'
+expand_limits_scale <- function(scale, expansion = expand_scale(0, 0), limits = waiver(),
+                                coord_limits = NULL) {
+  limits <- limits %|W|% scale$get_limits()
+
+  if (scale$is_discrete()) {
+    coord_limits <- coord_limits %||% c(NA_real_, NA_real_)
+    expand_limits_discrete(
+      limits,
+      expansion,
+      coord_limits,
+      range_continuous = scale$range_c$range
+    )
+  } else {
+    # using the inverse transform to resolve the NA value is needed for date/datetime/time
+    # scales, which refuse to transform objects of the incorrect type
+    coord_limits <- coord_limits %||% scale$trans$inverse(c(NA_real_, NA_real_))
+    coord_limits_scale <- scale$trans$transform(coord_limits)
+    expand_limits_continuous(limits, expansion, coord_limits_scale)
+  }
+}
+
+expand_limits_continuous <- function(limits, expansion = expand_scale(0, 0), coord_limits = c(NA, NA)) {
```

Comment:
I assume most of this code has been moved from elsewhere. Is there anything I should read closely?